### PR TITLE
Csr2bsr

### PR DESCRIPTION
**Pull request by jsandham**
Adding csr2bsr and bsr2csr for hipsparse

### CODE REVIEW

File: library/include/hipsparse.h [hunk 1]
```diff
@@ -1872,6 +1872,21 @@ hipsparseStatus_t hipsparseZnnz(hipsparseHandle_t         handle,
                                 int*                      nnzPerRowColumn,
                                 int*                      nnzTotalDevHostPtr);
 
+/* Description: 
+   This function computes the number of nonzero elements per row or column and the total number of nonzero elements in a dense matrix. */
```
Comment (jsandham) [inline library/include/hipsparse.h]: Not to self. Fix this comment

File: library/src/nvcc_detail/hipsparse.cpp [hunk 2]
```diff
@@ -3847,6 +3847,238 @@ hipsparseStatus_t hipsparseZcsr2hyb(hipsparseHandle_t         handle,
...
+                                    hipComplex*                bsrValC,
+                                    int*                       bsrRowPtrC,
+                                    int*                       bsrColIndC)
+{
+    return hipCUSPARSEStatusToHIPStatus(cusparseCcsr2bsr((cusparseHandle_t)handle,
+                                                         hipDirectionToCudaDirection(dirA),
+                                                         m,
+                                                         n,
+                                                         (const cusparseMatDescr_t)descrA,
+                                                         (cuComplex*)csrValA,
```
Comment (jsandham) [inline library/src/nvcc_detail/hipsparse.cpp]: Not to self should be (const cuComplex*)csrValA

File: library/src/nvcc_detail/hipsparse.cpp [hunk 3]
```diff
@@ -3847,6 +3847,238 @@ hipsparseStatus_t hipsparseZcsr2hyb(hipsparseHandle_t         handle,
...
+                                    hipDoubleComplex*          bsrValC,
+                                    int*                       bsrRowPtrC,
+                                    int*                       bsrColIndC)
+{
+    return hipCUSPARSEStatusToHIPStatus(cusparseZcsr2bsr((cusparseHandle_t)handle,
+                                                         hipDirectionToCudaDirection(dirA),
+                                                         m,
+                                                         n,
+                                                         (const cusparseMatDescr_t)descrA,
+                                                         (cuDoubleComplex*)csrValA,
```
Comment (jsandham) [inline library/src/nvcc_detail/hipsparse.cpp]: Not to self should be (const cuDoubleComplex*)csrValA

File: library/src/nvcc_detail/hipsparse.cpp [hunk 4]
```diff
@@ -3847,6 +3847,238 @@ hipsparseStatus_t hipsparseZcsr2hyb(hipsparseHandle_t         handle,
...
+                                    hipComplex*                csrValC,
+                                    int*                       csrRowPtrC,
+                                    int*                       csrColIndC)
+{
+    return hipCUSPARSEStatusToHIPStatus(cusparseCbsr2csr((cusparseHandle_t)handle,
+                                                         hipDirectionToCudaDirection(dirA),
+                                                         mb,
+                                                         nb,
+                                                         (const cusparseMatDescr_t)descrA,
+                                                         (cuComplex*)bsrValA,
```
Comment (jsandham) [inline library/src/nvcc_detail/hipsparse.cpp]: Not to self should be (const cuComplex*)bsrValA

File: library/src/nvcc_detail/hipsparse.cpp [hunk 5]
```diff
@@ -3847,6 +3847,238 @@ hipsparseStatus_t hipsparseZcsr2hyb(hipsparseHandle_t         handle,
...
+                                    hipDoubleComplex*          csrValC,
+                                    int*                       csrRowPtrC,
+                                    int*                       csrColIndC)
+{
+    return hipCUSPARSEStatusToHIPStatus(cusparseZbsr2csr((cusparseHandle_t)handle,
+                                                         hipDirectionToCudaDirection(dirA),
+                                                         mb,
+                                                         nb,
+                                                         (const cusparseMatDescr_t)descrA,
+                                                         (cuDoubleComplex*)bsrValA,
```
Comment (jsandham) [inline library/src/nvcc_detail/hipsparse.cpp]: Not to self should be (const cuDoubleComplex*)bsrValA

File: library/src/nvcc_detail/hipsparse.cpp [hunk 6]
```diff
@@ -3847,6 +3847,238 @@ hipsparseStatus_t hipsparseZcsr2hyb(hipsparseHandle_t         handle,
                          hipHybPartitionToCudaHybPartition(partitionType)));
```
Comment (jsandham) [inline library/src/nvcc_detail/hipsparse.cpp]: Not to self. Missing csr2bsr_nnz for nvcc!!!

File: clients/include/testing_csr2bsr.hpp [hunk 7]
```diff
@@ -0,0 +1,703 @@
...
+                                       m,
+                                       n,
+                                       csr_descr,
+                                       dcsr_row_ptr,
+                                       dcsr_col_ind,
+                                       block_dim,
+                                       bsr_descr,
+                                       dbsr_row_ptr,
+                                       &bsr_nnzb);
+
```
Comment (jsandham) [inline clients/include/testing_csr2bsr.hpp]: Note to self. Add verify hipsparse status after csr2bsr_nnz call as well:

if(m < 0 || n < 0 || block_dim < 0)
        {
            verify_hipsparse_status_invalid_size(status, "Error: m < 0 || n < 0 || block_dim < 0");
        }
        else
        {
            verify_hipsparse_status_success(status, "m >= 0 && n >= 0 && block_dim >= 0");
        }

File: clients/include/testing_csr2bsr.hpp [hunk 8]
```diff
@@ -0,0 +1,703 @@
...
+    CHECK_HIP_ERROR(
+        hipMemcpy(dcsr_row_ptr, hcsr_row_ptr.data(), sizeof(int) * (m + 1), hipMemcpyHostToDevice));
+    CHECK_HIP_ERROR(
+        hipMemcpy(dcsr_col_ind, hcsr_col_ind.data(), sizeof(int) * nnz, hipMemcpyHostToDevice));
+    CHECK_HIP_ERROR(hipMemcpy(dcsr_val, hcsr_val.data(), sizeof(T) * nnz, hipMemcpyHostToDevice));
+
+    if(argus.unit_check)
+    {
+        // Obtain BSR nnzb
+        int bsr_nnzb;
```
Comment (jsandham) [inline clients/include/testing_csr2bsr.hpp]: Note to self. Should also check using device pointer for bsr_nnzb as well

File: library/include/hipsparse.h [hunk 9]
```diff
@@ -1872,6 +1872,21 @@ hipsparseStatus_t hipsparseZnnz(hipsparseHandle_t         handle,
...
+HIPSPARSE_EXPORT
+hipsparseStatus_t hipsparseXcsr2bsr_nnz(hipsparseHandle_t          handle,
```
Comment (ntrost57) [inline library/include/hipsparse.h]: cusparse API is  `cusparseXcsr2bsrNnz`

**Comment by ntrost57** [comment]
Please also test nvidia path of hipsparse (Saad can grant you access to our V100 machines)

File: library/include/hipsparse.h (hunk 1 above)
Comment (jsandham) [reply library/include/hipsparse.h]: Fixed

File: library/src/nvcc_detail/hipsparse.cpp (hunk 3 above)
Comment (jsandham) [reply library/src/nvcc_detail/hipsparse.cpp]: FIxed

File: library/src/nvcc_detail/hipsparse.cpp (hunk 4 above)
Comment (jsandham) [reply library/src/nvcc_detail/hipsparse.cpp]: Fixed

File: library/src/nvcc_detail/hipsparse.cpp (hunk 6 above)
Comment (jsandham) [reply library/src/nvcc_detail/hipsparse.cpp:3872]: Fixed

File: library/include/hipsparse.h (hunk 9 above)
Comment (jsandham) [reply library/include/hipsparse.h]: Fixed in latest push

File: library/src/nvcc_detail/hipsparse.cpp [hunk 10]
```diff
@@ -3628,6 +3628,31 @@ hipsparseStatus_t hipsparseZnnz(hipsparseHandle_t         handle,
...
+                                       int                        n,
+                                       const hipsparseMatDescr_t  descrA,
+                                       const int*                 csrRowPtrA,
+                                       const int*                 csrColIndA,
+                                       int                        blockDim,
+                                       const hipsparseMatDescr_t  descrC,
+                                       int*                       bsrRowPtrC,
+                                       int*                       bsrNnzb)
+{
+    return hipCUSPARSEStatusToHIPStatus(cusparseScsr2bsrNnz((cusparseHandle_t)handle,
```
Comment (ntrost57) [inline library/src/nvcc_detail/hipsparse.cpp:3643]: should be `cusparseXcsr2bsrNnz()`

File: clients/include/utility.hpp [hunk 11]
```diff
@@ -968,6 +969,269 @@ inline void host_csr_to_csc(int                     M,
...
+#endif
+        for(size_t i = 0; i < csr_row_ptr.size(); i++)
+        {
+            bsr_row_ptr[i] = (csr_row_ptr[i] - csr_base) + bsr_base;
+        }
+
+        nnzb = bsr_row_ptr[mb] - bsr_row_ptr[0];
+
+        bsr_col_ind.resize(nnzb, 0);
+        bsr_val.resize(nnzb * block_dim * block_dim, static_cast<T>(0));
```
Comment (ntrost57) [inline clients/include/utility.hpp:1005]: need to change `static_cast<T>(0)` with `make_DataType<T>(0)` else nvcc will fail

File: clients/include/utility.hpp [hunk 12]
```diff
@@ -968,6 +969,269 @@ inline void host_csr_to_csc(int                     M,
...
+    for(int i = 0; i < mb; i++)
+    {
+        bsr_row_ptr[i + 1] += bsr_row_ptr[i];
+    }
+
+    nnzb = bsr_row_ptr[mb] - bsr_row_ptr[0];
+
+    // find bsr col indices array
+    bsr_col_ind.resize(nnzb, 0);
+    bsr_val.resize(nnzb * block_dim * block_dim, static_cast<T>(0));
```
Comment (ntrost57) [inline clients/include/utility.hpp:1062]: need to change `static_cast<T>(0)` with `make_DataType<T>(0)` else nvcc will fail

File: clients/include/utility.hpp [hunk 13]
```diff
@@ -968,6 +969,269 @@ inline void host_csr_to_csc(int                     M,
...
+                            std::vector<int>&       csr_col_ind,
+                            std::vector<T>&         csr_val)
+{
+    int m    = Mb * block_dim;
+    int n    = Nb * block_dim;
+    int nnzb = bsr_row_ptr[Mb] - bsr_row_ptr[0];
+
+    csr_row_ptr.resize(m + 1, 0);
+    csr_col_ind.resize(nnzb * block_dim * block_dim, 0);
+    csr_val.resize(nnzb * block_dim * block_dim, static_cast<T>(0));
```
Comment (ntrost57) [inline clients/include/utility.hpp:1154]: need to change `static_cast<T>(0)` with `make_DataType<T>(0)` else nvcc will fail

**Pull request merged**